src/data_provider.rs

```diff
@@ -1,174 +1,267 @@
 use std::thread::{self, JoinHandle};
-use std::sync::{Arc, Mutex, RwLock, mpsc::{self, Sender, Receiver}, atomic::{AtomicBool, Ordering}};
-use tokio::time::{sleep, Duration, timeout};
-use tokio::runtime::Runtime;
+use std::time::Duration;
+
+use gio::ListStore;
+use glib::Object;
+use gtk::{self, gio, glib, prelude::*};
 
+use async_channel::Sender;
 use givc_client::{self, AdminClient};
-use givc_common::query::{QueryResult, Event, VMStatus, TrustLevel};
+use givc_common::query::{Event, TrustLevel, VMStatus};
+use tokio::runtime::Builder;
 
+use crate::data_gobject::DataGObject;
 use crate::vm_gobject::VMGObject;
 //use crate::settings_gobject::SettingsGObject;//will be in use in the future
+
 use crate::{ADMIN_SERVICE_ADDR, ADMIN_SERVICE_PORT};
 
 pub mod imp {
     use super::*;
 
+    #[derive(Debug, Clone)]
+    pub struct TypedListStore<T: IsA<Object>>(ListStore, std::marker::PhantomData<T>);
+
+    pub struct LanguageRegionEntry {
+        pub code: String,
+        pub display: String,
+    }
+
+    impl<T: Into<String>, U: Into<String>> From<(T, U)> for LanguageRegionEntry {
+        fn from(val: (T, U)) -> Self {
+            Self {
+                code: val.0.into(),
+                display: val.1.into(),
+            }
+        }
+    }
+
+    impl From<LanguageRegionEntry> for DataGObject {
+        fn from(val: LanguageRegionEntry) -> Self {
+            Self::new(val.code, val.display)
+        }
+    }
+
+    pub struct LanguageRegionData {
+        pub timezones: Vec<LanguageRegionEntry>,
+        pub current_timezone: Option<String>,
+        pub languages: Vec<LanguageRegionEntry>,
+        pub current_language: Option<String>,
+    }
+
+    impl<T: IsA<Object>> TypedListStore<T> {
+        pub fn new() -> Self {
+            Self(ListStore::new::<T>(), std::marker::PhantomData)
+        }
+
+        pub fn get(&self, index: u32) -> Option<T> {
+            self.item(index).and_then(|item| item.downcast().ok())
+        }
+
+        pub fn iter(&self) -> impl Iterator<Item = T> {
+            let s = self.clone();
+            (0..).map_while(move |idx| s.get(idx))
+        }
+    }
+
+    impl<T: IsA<Object>, L: IsA<ListStore>> From<L> for TypedListStore<T> {
+        fn from(store: L) -> Self {
+            Self(store.upcast(), std::marker::PhantomData)
+        }
+    }
+
+    impl<T: IsA<Object>> Deref for TypedListStore<T> {
+        type Target = ListStore;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    type Task = Box<
+        dyn for<'a> FnOnce(
+                RwLockReadGuard<'a, AdminClient>,
+            ) -> std::pin::Pin<
+                Box<dyn std::future::Future<Output = Result<(), String>> + 'a>,
+            > + Sync
+            + Send,
+    >;
+
+    type TaskSender = Sender<(Task, Sender<Result<(), String>>)>;
+
     #[derive(Debug)]
     pub struct DataProvider {
-        store: Arc<Mutex<ListStore>>,
+        store: TypedListStore<VMGObject>,
```

Comment:
Have you tested it on ghaf? Is list updating successfully ?

Reply:
I'll run a few more tests to verify properly; however, as `ListStore` is a `GObject`, it already has internal reference counting, adding external one shouldn't affect things.

Reply:
The list does update successfully; I did a little change to remove items from the list on shutdown to make it more apparent. The commit can be easily dropped.

Reply:
If is removed, user won't be able to start it again, cause there will be no item in the view. Please, revert.